rfcs/0002-assetid.md

```diff
@@ -0,0 +1,107 @@
+FilePath - The current way of referencing assets. It will still be supported in scenarios where it is possible.
+URI - A way for users to plug in their own custom asset referencing scheme along with their own resolving systems.
+
+I expect users that write code to primarily reference assets using the FilePath variant. References in data, for example those that are created by a Prefab editor when referencing meshes or textures would use UUID references. This is closer to the expectation of the user since the user did not specify a specific path, but a reference to an asset. 
+
+If a user created a prefab using a prefab editor and references an asset in a file that was later renamed or moved, it would be unexpected for the reference to no longer resolve since the user did not specify a path in the prefab editor. In the case of loading an asset by file path in code, it would be be expected behaviour for the load to fail if a file is removed however.
+
+URI is a general way of referencing an asset that does not exist in the filesystem and was not imported by the asset pipeline. An example of this could be an asset stored in a database somewhere that is specific to the game it is resolved in. URIs are included in AssetID as a point of further extension for game developers and will not be used by any of the initial Asset Pipeline systems.
+The intention is to provide traits that can be implemented to make it easier to integrate custom IDs into the asset loading process.
+
+# Reference-Level Explanation
+[reference-level-explanation]: #reference-level-explanation
+
+I propose to define the following in amethyst::assets.
+```
+pub type AssetUUID = [u8; 16];
+#[derive(Clone, Serialize, Deserialize, Hash)]
+pub enum AssetID {
+    UUID(AssetUUID),
+    FilePath(PathBuf),
+    URI(URI),
+}
+```
+
+When an asset is imported from a file, an AssetUUID is generated for it and maintained by the Asset Hub daemon in metadata (see other, future RFC).
+
+Once the Asset Pipeline proposal is fully realized, the AssetID will be used in various parts of the engine/asset pipeline. I don't believe there is value in implementing it before other parts of the Asset Pipeline is ready.
+
+### Import
+Each source file can be imported independently in parallel. When importing a single source file, possibly multiple resulting assets will be assigned an AssetUUID each and they may have AssetID dependencies of three kinds: Build, load and instantiate.
+In some cases these dependencies will be defined with the FilePath variant. After the import completes but before persisting metadata, the Asset Pipeline will attempt to resolve the FilePath dependencies of these assets to their AssetUUID equivalents. If the path cannot be resolved, the import will result in an error. This will ensure that the Asset Pipeline and all tools only have to handle stable AssetUUID references that map to specific assets regardless of the filesystem state.
```

Comment:
> In some cases these dependencies will be defined with the FilePath variant

This sentence is contradictory to the rest of this paragraph. I was actually a bit surprised until i've read the rest of it. I'd actually avoid the `FilePath` name specifically here, as it might be confusing. There is nothing special about filesystem paths really.

How about that?

> In some cases, the asset source file might specify dependencies in any way, e.g. relative paths or other format-specific methods. After the import completes but before persisting metadata, the Asset Pipeline will attempt to resolve the dependencies of these assets to their AssetUUID equivalents in a format-specific way. If the dependency cannot be resolved, the import will result in an error.